### PR TITLE
feat(19) shelf invitations

### PR DESCRIPTION
**Pull request by dubbha**
**server**:
 - API: shelfUserInvite, shelfUserInviteAccept, shelfUserDelete
 - Entities: Shelf, ShelfInvitation, UserToShelf, Flower
 - DB service: getShelfInvitationByToken, saveShelfInvitation, deleteShelfInvitation, getShelfInvitationsByUserEmail, getUserToShelf, saveUserToShelf, deleteUserToShelf, getShelfById, getFlowerById, tests
 - shelfAdminMiddleware checking the user is authorized to perform shelf admin operations
 - ShelfController
 - ShelfService
 - Invitation emails for both existing and unknown users

**web**:
 - shelf state, actions, reducer, sagas (invite, inviteAccept, deleteUser)
 - ShelfInvitationAccept component
- InviteUserToShelf and DeleteUserFromShelf modal components

**BREAKING CHANGE**:
`SHELF_INVITATION_EXPIRY_TIME` is now required in `.env` of `server`, set to 720 hours (30 days):
```
SHELF_INVITATION_EXPIRY_TIME=720
```

### CODE REVIEW

**Review by karolinayanko — APPROVED** [review]

File: packages/server/src/db/db.service.ts [hunk 1]
```diff
@@ -21,13 +27,21 @@ export default class DBService {
...
   constructor() {
     this.userRepository = getRepository(UserEntity);
     this.emailResetRepository = getRepository(EmailResetEntity);
     this.emailVerificationRepository = getRepository(EmailVerificationEntity);
     this.passwordResetRepository = getRepository(PasswordResetEntity);
     this.socialProfileRepository = getRepository(SocialProfileEntity);
+    this.shelfRepository = getRepository(ShelfEntity);
+    this.flowerRepository = getRepository(FlowerEntity);
+    this.shelfInvitationRepository = getRepository(ShelfInvitationEntity);
+    this.userToShelfRepository = getRepository(UserToShelfEntity);
```
Comment (dartwlad) [inline packages/server/src/db/db.service.ts:46]: I would suggest splitting this service, in scope of future PR's.
src/db/auth.service
src/db/user-management.service
src/db/shelf.service
src/db/flower.service
Or create some kind of CRUD DB service.

File: packages/server/src/services/email-sending.service.ts [hunk 2]
```diff
@@ -63,4 +67,34 @@ export default class EmailSendingService {
...
   }
+
+  public async sendShelfInvitationMessageUnknownUser(
+    userEmail: string,
+    host: string,
+  ): Promise<void> {
+    const link = `${host}${Pages.SIGNUP_PAGE}`;
+    const mailOptions = {
+      to: userEmail,
+      subject: 'You have been invited to a take care of a flower shelf',
```
Comment (dartwlad) [inline packages/server/src/services/email-sending.service.ts]: You have been invited to take care of a flower shelf

File: packages/server/src/services/email-sending.service.ts [hunk 3]
```diff
@@ -63,4 +67,34 @@ export default class EmailSendingService {
...
+  public async sendShelfInvitationMessage(
+    userEmail: string,
+    name: string,
+    host: string,
+    shelfInvitationToken: string
+  ): Promise<void> {
+    const link = `${host}${Pages.SHELF_INVITATION_ACCEPT_PAGE}?${QueryParams.SHELF_INVITATION_TOKEN}=${shelfInvitationToken}`;
+    const mailOptions = {
+      to: userEmail,
+      subject: 'You have been invited to a take care of a flower shelf',
```
Comment (dartwlad) [inline packages/server/src/services/email-sending.service.ts]: Same

File: packages/web/src/components/Shelves/AddNewShelfForm/AddNewShelfForm.tsx [hunk 4]
```diff
@@ -1,5 +1,7 @@
 import React, { useState, FormEvent, useEffect } from 'react';
 import { Form, SubmitButton } from 'elements';
+// import { InviteUserToShelf } from '../InviteUserToShelf';
+// import { DeleteUserFromShelf } from '../DeleteUserFromShelf';
```
Comment (dartwlad) [inline packages/web/src/components/Shelves/AddNewShelfForm/AddNewShelfForm.tsx]: Don't forget to remove it.

File: packages/web/src/components/Shelves/AddNewShelfForm/AddNewShelfForm.tsx [hunk 5]
```diff
@@ -128,6 +130,8 @@ export const AddNewShelfForm = ({ onSubmit }: Props) => {
         <div className="spinner-container" />
         <span>Create</span>
       </SubmitButton>
+      {/* <InviteUserToShelf shelfId={2} />
+      <DeleteUserFromShelf shelfId={2} userId={13} /> */}
```
Comment (dartwlad) [inline packages/web/src/components/Shelves/AddNewShelfForm/AddNewShelfForm.tsx]: Same

**Review by belgithub — APPROVED** [review]

**Review by dartwlad — APPROVED** [review]

File: packages/server/src/db/db.service.ts (hunk 1 above)
Comment (dubbha) [reply packages/server/src/db/db.service.ts:46]: Agree, it asks for refactoring, sepecially the test part.

**Pull request merged**